src/frontend/hooks/server/media.ts

```diff
@@ -1,52 +1,99 @@
 import {Observation} from '@mapeo/schema';
 import {BlobVariant} from '@mapeo/core/dist/types';
-import {useMutation, useQuery} from '@tanstack/react-query';
+import {useMutation, useQueries} from '@tanstack/react-query';
+import {SetRequired} from 'type-fest';
+import {URL} from 'react-native-url-polyfill';
 
-import {Photo} from '../../contexts/PhotoPromiseContext/types';
+import {DraftPhoto} from '../../contexts/PhotoPromiseContext/types';
 import {useProject} from './projects';
 
+type SavablePhoto = SetRequired<
+  Pick<DraftPhoto, 'originalUri' | 'previewUri' | 'thumbnailUri'>,
+  'originalUri'
+>;
+
 export function useCreateAttachmentsMutation() {
   const project = useProject();
 
   return useMutation({
-    mutationFn: async (photos: Photo[]) => {
+    mutationFn: async (photos: SavablePhoto[]) => {
       if (!project) throw new Error('Project instance does not exist');
 
       return Promise.all(
```

Comment:
yeah that makes sense. how does the retry stuff translate code-wise? is there something built into react query to handle the case of "we failed to save 1 of 3 photos, retry for that 1", or do we have to do something manually e.g. re-attempt a mutation in the `onSuccess` callback?

Reply:
part of me is wondering if this should be updated to only handle a single photo and then the consumer can decide to make concurrent calls if needing to save several photos.

not sure if there's an idiomatic approach that react-query suggests when it comes to saving "batches" of data, but my intuition is that mutations generally hold up better when they map 1-to-1 with the api that you're interacting with. for example, `project.$blobs.create()` accepts a single photo as opposed to many photos, so it feels more natural for the mutation to look similar. all just pure feeling though...

would updating this mutation so that it accepts a single photo and then calling several mutations in parallel make sense, especially given what you described? for example, after pressing the button to save the observation, saving the attachments could look like this:

```ts
const photosToSave = []

// Can also do Promise.all() but wrapped with a try/catch to manually handle the errors since react query won't in that case
const results = await Promise.allSettled(photosToSave.map((photo) => {
  return mutation.mutateAsync(photo))
})

// Depending on if we configure the `retry` option when creating the mutation, we could do checks on `results`

```

Reply:
more of a UX question, but what should happen in the case that saving some (or all) of the attachments doesn't work, even after the retries? should you still be allowed to save the observation, but only with the attachments that successfully saved? (e.g. observation may be missing some that you took)

doesn't seem like we do much safeguarding in Mapeo Mobile (https://github.com/digidem/mapeo-mobile/blob/develop/src/frontend/hooks/useDraftObservation.ts#L272), so probably not a good reference for this question 😄

Reply:
summarizing the problem that you've highlighted:

> Currently, we abort the process of saving an observation if saving any number of photos fails to save,
but this approach is prone to creating "orphaned" blobs.
> The alternative is to save the observation but excluding photos that failed to save, which is prone to an odd UX of an observation "missing" some attachments.
> This could potentially be alleviated by a more granular and informative UI about the photo-saving state, but currently there is nothing in place.
> Basically, which is worse: orphaned attachments or saving observations that seem to be missing attachments?

from just basic testing between the two paths, i actually found the case of "missing" attachments to be much more jarring, although that's mostly because of a lack of UI to surface that prior to saving the observation. 

For the sake of consistency and lack of a simple solution, I'll have the implementation maintain the same behavior as Mapeo Mobile and have the "orphaned" attachment problem, at least temporarily. Wondering if there's maybe some backend work that we can do to make this API usage less prone to such issues.

Reply:
re: batching mutations, just realized that if i go with the approach of a mutation per photo, i'll have to do more work to get the pending state of "saving all photos". the `isPending` from `useMutation()` will update for each concurrent mutate call I make. seems like i'd have to use [`useIsMutating()`](https://tanstack.com/query/latest/docs/framework/react/reference/useIsMutating) or [`useMutationState()`](https://tanstack.com/query/latest/docs/framework/react/reference/useMutationState) in order to achieve that 🤔

Reply:
(somewhat) addressed via https://github.com/digidem/CoMapeo-mobile/pull/187/commits/db95f650e6c96024644aa57811fdf0d726b0b55e